### ISSUE TITLE
Model with save_resume gives worse model skill than without save_resume

### ISSUE THREAD

**Issue by whitish** [VowpalWabbit/vowpal_wabbit#1958]
#### Describe the bug
I am trying to implement continuous training with VW (using imbalanced data). With and without _--save_resume_ I expect the same AUC, logloss etc for the first batch of the training data. And for the following iterations I expect that with _--save_resume_ metrics will be better. However, even after the first training iteration I have got different results. For example:

#### train/test on batch 1 (positive event / all events: 177/193028)
`hadoop fs -cat train_batch_1.vw |  vw --cache_file model.cache --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet --final_regressor model.model`

`hadoop fs -cat test_batch_1.vw | vw --loss_function logistic --link logistic --quiet -t -i model.model -p test_batch_1.pred`
##### result without --save_resume
`roc-auc: 0.8851, logloss: 0.0063, logloss_baseline: 0.0073, logloss / baseline: 0.8549, calibration: 1.0811
`
##### result with --save_resume (training command is the same but with --save_resume flag; testing command is exactly the same)
`roc-auc: 0.8308, logloss: 0.1278, logloss_baseline: 0.0073, logloss / baseline: 17.436, calibration: 105.9981`

At this iteration results should be identical, but they are not.

#### train/test on batch 2 (positive event / all events: 301/284814)
`hadoop fs -cat train_batch_2.vw |  vw --cache_file model.cache --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet --final_regressor model.model
`

`hadoop fs -cat test_batch_2.vw | vw --loss_function logistic --link logistic --quiet -t -i model.model -p test_batch_2.pred
`

##### result without --save_resume
`roc-auc: 0.8515, logloss: 0.0073, logloss_baseline: 0.0083, logloss / baseline: 0.8836, calibration: 0.9442
`
#### result with --save_resume (training command is the same but with --save_resume flag; testing command is exactly the same)
`roc-auc: 0.6121, logloss: 0.3341, logloss_baseline: 0.0083, logloss / baseline: 40.2587, calibration: 267.9847`

#### train/test on batch 3 (positive event / all events: 465/460322)
`hadoop fs -cat train_batch_3.vw |  vw --cache_file model.cache --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet --final_regressor model.model
`

`hadoop fs -cat test_batch_3.vw | vw --loss_function logistic --link logistic --quiet -t -i model.model -p test_batch_3.pred
`
##### result without --save_resume
`roc-auc: 0.8335, logloss: 0.0071, logloss_baseline: 0.008, logloss / baseline: 0.8937, calibration: 1.2761
`
#### result with --save_resume (training command is the same but with --save_resume flag; testing command is exactly the same)
`roc-auc: 0.67, logloss: 0.3511, logloss_baseline: 0.008, logloss / baseline: 44.014, calibration: 292.4441`

**Comment by arielf** [comment]
Thanks for the report.

One potential problem here is using 3 different inputs (`test_batch_<N>.vw`) but they are sharing the same cache file `model.cache` in multiple passes.

Since this is not a reproducible example, and a not so big data-set (< 1M examples)
In order to further narrow the issue it would be really helpful if you provide more info:

- Can you reproduce with local files passed directly (`-d ...`) instead of via hadoop/stdin and let `vw itself create the (separate) cache files per batch, deriving their names from the inputs (option: `-c`). This will ensure cache files are not mixed-up.
- If problem still exists after separating cache files, can you reproduce the bug without the `--ftrl` flags?

**Comment by whitish** [comment]
Replaced _--cache_file_ with _-c_ to be more confident about not mixing the cachesI've got exactly the same results (I am also using _--kill_cache_).

Output (with --ftrl):
```
# no --save_resume

# train/test iteration 1 (positive events / all events: 177/193028)
vw -d train_1.vw -c --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet --final_regressor model.model
vw -d test_1.vw --loss_function logistic --link logistic --quiet -t -p pred_1.pred

# train/test iteration 2 (positive events / all events: 301/284814)
vw -d train_2.vw -c --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet -i model.model --final_regressor model.model --kill_cache
vw -d test_2.vw --loss_function logistic --link logistic --quiet -t -p pred_2.pred


# with --save_resume

# train/test iteration 1 (positive events / all events: 177/193028)
vw -d train_1.vw -c --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet --final_regressor model.model --save_resume
vw -d test_1.vw --loss_function logistic --link logistic --quiet -t -i model.model -p pred_1.pred

# train/test iteration 2 (positive events / all events: 301/284814)
vw -d train_2.vw -c --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --quiet -i model.model --final_regressor model.model --kill_cache --save_resume
vw -d test_2.vw --loss_function logistic --link logistic --quiet -t -i model.model -p pred_2.pred


# results for the 1st and 2nd batches:
# no --save_resume
# roc-auc: 0.8851, pr_avg: 0.0061, brier: 0.0009, logloss: 0.0063, logloss_baseline: 0.0073, logloss / baseline: 0.8549, calibr: 1.0811
# roc-auc: 0.8515, pr_avg: 0.0074, brier: 0.0011, logloss: 0.0073, logloss_baseline: 0.0083, logloss / baseline: 0.8836, calibr: 0.9442

# with --save_resume
# results:
# roc-auc: 0.8308, pr_avg: 0.003, brier: 0.0385, logloss: 0.1278, logloss_baseline: 0.0073, logloss / baseline: 17.436, calibr: 105.9981
# roc-auc: 0.5637, pr_avg: 0.0012, brier: 0.0011, logloss: 0.013, logloss_baseline: 0.0083, logloss / baseline: 1.5678, calibr: 7.2466

```




Output (without --ftrl):
```
# no --save_resume

# train/test iteration 1 (positive events / all events: 177/193028)
vw -d train_1.vw -c --loss_function logistic --passes 4 --l1 1.0 --l2 0.1 --quiet --final_regressor model.model
vw -d test_1.vw --loss_function logistic --link logistic --quiet -t -p pred_1.pred

# train/test iteration 2 (positive events / all events: 301/284814)
vw -d train_2.vw -c --loss_function logistic --passes 4 --l1 1.0 --l2 0.1 --quiet -i model.model --final_regressor model.model --kill_cache
vw -d test_2.vw --loss_function logistic --link logistic --quiet -t -p pred_2.pred



# with --save_resume

# train/test iteration 1 (positive events / all events: 177/193028)
vw -d train_1.vw -c --loss_function logistic --passes 4 --l1 1.0 --l2 0.1 --quiet --final_regressor model.model --save_resume
vw -d test_1.vw --loss_function logistic --link logistic --quiet -t -i model.model -p pred_1.pred

# train/test iteration 2 (positive events / all events: 301/284814)
vw -d train_2.vw -c --loss_function logistic --passes 4 --l1 1.0 --l2 0.1 --quiet -i model.model --final_regressor model.model --kill_cache --save_resume
vw -d test_2.vw --loss_function logistic --link logistic --quiet -t -i model.model -p pred_2.pred

# results for the 1st and 2nd batches:
# no --save_resume
# roc-auc: 0.5, pr_avg: 0.0009, brier: 0.25, logloss: 0.6931, logloss_baseline: 0.0073, logloss / baseline: 94.5604, calibr: 545.2768
# roc-auc: 0.5, pr_avg: 0.0011, brier: 0.25, logloss: 0.6931, logloss_baseline: 0.0083, logloss / baseline: 83.53, calibr: 473.113

# with --save_resume
# roc-auc: 0.2206, pr_avg: 0.0034, brier: 0.2492, logloss: 0.6916, logloss_baseline: 0.0073, logloss / baseline: 94.3513, calibr: 544.4376
# roc-auc: 0.5, pr_avg: 0.0011, brier: 0.25, logloss: 0.6931, logloss_baseline: 0.0083, logloss / baseline: 83.53, calibr: 473.113

```

**Comment by arielf** [comment]
Thanks Aleksei

#### Let's focus on the 1st problem described: 

"even after the first training iteration I have got different results" (I'm guessing vs adding `--save_resume` to the command line)

I tried and am not able to reproduce this.

Can you provide a fully reproducible example? Both full script and data-set?
If the bug is there, it should manifest with any (much smaller) dataset.

#### Here's what I tried:

Use latest `vw`:  `--version: 8.7.0 `
I get the same results with and without `--save_resume`. 
Even when I use all the options in your example on one of the binary-classification data-sets in the `vw` source tree:

```
# Only remove --quiet, so we can see results:

% vw -d test/train-sets/rcv1_small.dat   -c --loss_function logistic --passes 4 --ftrl --l1 1.0 --l2 0.1 --ftrl_alpha 0.1 --ftrl_beta 1.0 --final_regressor model.model  [--save_resume]

# With or without the `--save_resume` option, I get exactly the same output (progress + final loss)
using l1 regularization = 1
using l2 regularization = 0.1
final_regressor = model.model
Enabling FTRL based optimization
Algorithm used: Proximal-FTRL
ftrl_alpha = 0.1
ftrl_beta = 1
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
using cache_file = test/train-sets/rcv1_small.dat.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.693147 0.693147            1            1.0  -1.0000   0.0000      128
0.693147 0.693147            2            2.0  -1.0000   0.0000       44
0.693147 0.693147            4            4.0  -1.0000   0.0000      190
0.694494 0.695841            8            8.0   1.0000  -0.0214       34
0.696218 0.697942           16           16.0   1.0000  -0.0209      193
0.695041 0.693865           32           32.0   1.0000  -0.0433       48
0.692895 0.690749           64           64.0  -1.0000  -0.0785      119
0.693881 0.694866          128          128.0  -1.0000  -0.0751       38
0.690374 0.686868          256          256.0  -1.0000  -0.1554       23
0.688114 0.685855          512          512.0  -1.0000  -0.1933       58
0.681730 0.681730         1024         1024.0   1.0000  -0.1033       35 h
0.668736 0.655855         2048         2048.0   1.0000  -0.1182       89 h

finished run
number of examples per pass = 900
passes used = 4
weighted example sum = 3600.000000
weighted label sum = -272.000000
average loss = 0.614572 h
best constant = -0.151400
best constant's loss = 0.690290
total feature number = 284052
```

The final models are different, as expected:
```
without `--save resume`: 11189 bytes
with `--save_resume`: 149597 bytes, (due to additional per feature learning-rates being saved.)
```

#### Two more unrelated things I noticed:

- For effective learning,  regularization values should probably be _much_ smaller than 1.0. `vw` applies them as-is directly to the sum-of-gradients, so a high value like `1.0` in binary classification, almost disables learning.
- When doing multiple passes, 1:10 automatic holdout automatically gets applied, not sure if this is what you want on such a small signal unbalanced data-set.  `--quiet` hides this (holdout) fact from you. You may want to add the `--holdout_off` option.

**Comment by JohnLangford** [comment]
Closing this as resolved by subsequent changes to VW's save_resume behavior.

## Analysis

We investigated the FTRL + `--save_resume` code path in detail:

1. **Weight storage is correct** — both the save_resume and non-save_resume paths store and load W_XT (the prediction weight) in the same slot. The save_resume path additionally persists W_ZT/W_G2 optimizer state into adjacent slots, but FTRL's predict function reads only `w[W_XT]`.

2. **L1/L2 lazy state (gravity/contraction) is unaffected** — FTRL applies L1/L2 regularization directly in its update formula (`inner_update_proximal`), not through GD's lazy gravity/contraction mechanism. These values stay at their defaults (0.0 and 1.0) during FTRL training.

3. **Prediction clamping (min_label/max_label) is identical** — logistic loss initializes these to -50/50, and the `set_minmax` callback can only widen the range during training.

4. **Multiple save_resume fixes landed between 8.7.0 and 9.0.0:**
   - `523c5e5` — make `--l1` and `--l2` work through save_resume (previously a known failure)
   - `379355e` — fix PRG state preservation in save_resume
   - `52649f5` (PR #3502) — make save_resume the **default** model format, removing the need for `--save_resume` flag entirely

5. **Save_resume is now the default** — since VW 9.0.0, all models are saved in resume format unless `--predict_only_model` is specified. The existing FTRL test suite (tests 70/71, 103-106) trains, saves, loads, and predicts through this default path and passes. If this bug still existed, it would affect all FTRL users.

We also note that in the second comment's reproduction script, the "no save_resume" test commands are missing `-i model.model` (present in the "with save_resume" commands), which would cause prediction with zero weights rather than the trained model.
